### PR TITLE
Improve devcontainers

### LINKED ISSUE #104: Make dev container setup a more complete development environment
This was discussed in #102 when adding initial dev container support for testing.

We could envisage having an entirely self-contained development environment for users who do not want to install OpenSAFELY tooling by hand. It reduces the installation process to:

1. Install Docker
2. Install Visual Studio Code
3. Clone a research repository
4. Open the research repository in Visual Studio Code, and build a dev container from the configuration.

This also could be used in Codespaces for users unable to install the OpenSAFELY tooling themselves.

The other advantage of this is it *should*[^1] be the same setup cross-platform, whether Windows/macOS/Linux/Codespaces.

***

For R, @remlapmot had [built a working R-in-dev-container-setup](https://github.com/opensafely/research-template/pull/102#issuecomment-1629339277), based on the OpenSAFELY R image.

Stata might be more of a problem.

[^1]: Assuming there are no behavioural quirks per platform, which is far from certain.

### PR DESCRIPTION
**Pull request by lucyb**
References https://github.com/opensafely-core/codespaces-initiative/issues/43 and https://github.com/opensafely-core/codespaces-initiative/issues/9

Fixes #104 and https://github.com/opensafely-core/codespaces-initiative/issues/43.

This will provide a development environment for researchers containing:

- Python 3.10
- The equivalent Python packages based on the Python Action image, including ehrql
- R 4.0.5
- RStudio
- The equivalent R packages based on the R Action image

This can be used in Codespaces or locally via devcontainers in VS Code.

### CODE REVIEW

**Comment by Jongmassey** [comment]
Looking at Tom P's comments in the other thread, I see that not copying across the renv from the OpenSAFELY R action image saves a bunch of space, but unless Tom updates his base image we're going to miss any updates to the OpenSAFELY R package list. Did the base rstudio image not work?

**Comment by Jongmassey** [comment]
The Dockerfile now references the base rocker image and in my testing works fine. Autocompletion in r-studio works and all the libraries in the renv are present.  Image size is now 5.44GB, which isn't _great_ but is better than 8GB. Need to double-check all things Python are fine

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    # Pip for Python 3.10 isn't included in deadsnakes, so install separately
+    curl https://bootstrap.pypa.io/get-pip.py | python3.10 &&\
+    # Set default python, so that the Python virtualenv works as expected
+    rm /usr/bin/python3 && ln -s /usr/bin/python3.10 /usr/bin/python3
+
+# install renv
+RUN --mount=type=cache,target=/cache,id=/cache-2004 R -e 'install.packages("renv", destdir="/cache"); renv::init(bare = TRUE)'
+
+# copy the renv directory from the OpenSAFELY R action image
+COPY --from=ghcr.io/opensafely-core/r /renv/ /renv/
```
Comment (lucyb) [inline .devcontainer/Dockerfile:25]:
```suggestion
COPY --from=ghcr.io/opensafely-core/r:latest /renv/ /renv/
```

Comment (lucyb) [reply .devcontainer/Dockerfile:25]: I imagine we'll move away from using the "latest" version of the r image at some point, so this just makes it more obvious what it's doing.

**Comment by remlapmot** [comment]
Just checking - are you sure that renv is configured correctly in this updated version? 

On starting an RStudio session I think I see only 32 R packages. If I then run `renv::activate()` I get 30 packages; whereas the current magic number for the r image is 438 packages.

Screenshot 2024-04-28 at 20 45 56

In `/renv/lib/R-4.0/x86_64-pc-linux-gnu` I see 422 packages (last time I saw 424 - not sure why 2 fewer this time) but I think you are still missing the `/usr/lib/R/library` packages from the r image (which `renv::activate()` needs to find). 

More simply you could try copying the packages from the r image into the system R package directories, such as `/usr/local/lib/R/site-library` (then there'd be no need to worry about `renv::activate()`-ing. I would probably clear out the system package directory first as there are packages in it from rocker). I admit that when I made my Dockerfile I didn't realise it is possible to copy files between images like this.

File: .devcontainer/postCreate.sh [hunk 2]
```diff
@@ -9,6 +9,12 @@ pip3 install --user -r .devcontainer/requirements.in
 #set RStudio working directory
 ! grep -q `pwd` ~/.config/rstudio/rstudio-prefs.json && cat ~/.config/rstudio/rstudio-prefs.json | jq ". + {\"initial_working_directory\":\"`pwd`\"}" >  ~/.config/rstudio/rstudio-prefs.json 
 
+#disable rstudio server auth
+! grep -q "auth-none=1" /etc/rstudio/rserver.conf && sudo tee -a /etc/rstudio/rserver.conf <<< "auth-none=1"
+
+#set default rstudio user
+! grep -q "USER=rstudio" /etc/environment && sudo tee -a /etc/environment <<< "USER=rstudio"
```
Comment (lucyb) [inline .devcontainer/postCreate.sh:16]: Is this something we could move to the Dockerfile instead?

Comment (Jongmassey) [reply .devcontainer/postCreate.sh:16]: Good point, yes we could and should. I'll do that now

Comment (Jongmassey) [reply .devcontainer/postCreate.sh]: Done

**Review by iaindillingham — APPROVED** [review]

**Pull request merged**